### ISSUE TITLE
(aws-s3): CDK in parallel mode. Lambda service: Could not unzip uploaded file.

### ISSUE THREAD

**Issue by heikkis** [aws/aws-cdk#19695]
### Describe the bug

I run CD pipeline in Github Actions which uses native AWS CDK parallel deployment.
Once in a while S3 asset uploading zip file cannot be opened by AWS (or MacOS etc.).

Could not unzip uploaded file. Please check your file, then try to upload again.

One case the file size was 203 Bytes. Filename of asset did not match either to md5sum or sha1 sum. Sample availible if someone from AWS can reach me out.

After deleting the zip file from CDK asset S3 bucket and doing the same deploy again, the file size was 1,6KB and it worked.

### Expected Behavior

No errors and parallel deployment works fine.

### Current Behavior

Zip file is corrupted. I cannot open the file at my machine even.

```
XXXXX | 4:09:34 PM | UPDATE_FAILED        | AWS::Lambda::Function                       | rds-cluster/XXXXX-XXXX-lambda (XXXXXX) Resource handler returned message: "Could not unzip uploaded file. Please check your file, then try to upload again. (Service: Lambda, Status Code: 400, Request ID: XXXXX-XXXXX-4c03-91d5-XXXXXXX)" (RequestToken: XXXXX-XXXXXX-XXXXX-9e9a-XXXXXX, HandlerErrorCode: InvalidRequest)
```

### Reproduction Steps

Not easy job to reproduce.

1. CDK parallel/concurrent deploy
2. Happens sometimes

### Possible Solution

.

### CDK CLI Version

2.39.1

### Node.js Version

16.4.12

### OS

Github managed runner

### Language

Typescript

### Language Version

Typescript (4.6.3)

**Comment by NGL321** [comment]
Since this is a super edge case (and likely to be troublesome to find the root cause), I am marking it as low priority for a fix. However, if you are able to identify the cause directly I can upgrade it's priority for a fix.

**Comment by tmokmss** [comment]
It might be related to this https://github.com/aws/aws-cdk/issues/9545. iirc `cdk deploy` will synthesize your app again on each invoke so your zip assets can be overwritten while being uploaded.
You can try to use different output directories per `cdk deploy` command with `--output` flag.

**Comment by heikkis** [comment]
@tmokmss synthetization is done only once. And CDK deploy command is using synth output folder as an application, it is a special way to execute deploy without having any synthetization happening any more.

**Comment by royby-cyberark** [comment]
I am also facing the same issue, investigated a bit, and I believe it is something in the way/place that the Asset zip is created. 

My reasoning is:
I run 4 concurrent cdk deploy --output <different folder for each run>
so my output folders are unique - but with the same asset.x subfolders.

I still get the error, which, suggests that when creating the zip, it is done in the same temp folder, regardless of the different output folders. which causes it to be corrupted for some. 

@heikkis - if you control the Lambda creation (is it a lambda?) then you can set the hash of the AssetCode per run, which will make it unique and so avoid the collision. 
https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_lambda/AssetCode.html

for me, I am getting this when deploying a custom resource Provider and so I can't manipulate its internal lambda asset. 
so i'm still trying to understand what to do.

**Comment by royby-cyberark** [comment]
@NGL321 i'm trying to identify the root cause, can you please help me by helping me find the place in code that the zip from the asset is created? 
I think that the reason the zip is corrupted is because it uses the same temp location for this, regardless of having different out folders.

Thanks!

**Comment by thdxr** [comment]
We are also running into this issue at SST since we deploy stacks in parallel by default and all stacks have an identical function within them.  I also believe this is a tmp file issue and not an s3 issue

**Comment by kichik** [comment]
FWIW I believe CDK Pipelines avoids this issue by separating the synth, asset push and deployment stages. It uses `cdk synth` to synth, `cdk-assets` cli to push assets, and then just normal CodePipeline actions for CloudFormation to update the stacks. I'm not sure what it does when using GitHub Actions instead of CodePipeline. But this way you don't have two processes uploading to S3 at once, and you don't have two processes writing the asset to disk at once or using a temporary folder. So whatever causes this issue, using three stages helps.

**Comment by heikkis** [comment]
Same issue with still exists when heavily using AWS CDK concurrent mode with stacks that required to upload assets, eg. AWS CDK custom resource build in lambda functions. 

The file is in the S3 but the size was only 203 bytes. It is AWS CDK log rotation custom lambda.

I have example file that I can pass privately to AWS if needed.


```
MacOS> unzip eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip
Archive:  eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip or
        eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip.zip, and cannot find eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip.ZIP, period.

MacOS > md5 eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip
MD5 (eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip) = bbb6bdc10dc6e66e7cca3f31788aa167

MacOS > hasum eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8
.zip 
67f79c2f10c016f71270ff963c16cd2997d62bae  eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip
```
